Google.PowerShell/Common/ActiveUserToken.cs

```diff
@@ -12,37 +14,131 @@ public class ActiveUserToken
+        /// </summary>
+        public ActiveUserToken(string userCredentialJson, Func<string> getActiveUser)
+        {
+            this.getActiveUser = getActiveUser;
+            activeUser = this.getActiveUser();
+
+            JToken parsedCredentialJson = JObject.Parse(userCredentialJson);
+            JToken accessTokenJson = parsedCredentialJson.SelectToken("access_token");
+
+            if (accessTokenJson == null || accessTokenJson.Type != JTokenType.String)
+            {
+                throw new InvalidDataException("Credential Json should contain access token key.");
+            }
+
+            AccessToken = accessTokenJson.Value<string>();
+
+            JToken tokenExpiryJson = parsedCredentialJson.SelectToken("token_expiry");
+
+            // Token from GCE does not have expiry.
+            if (tokenExpiryJson == null || tokenExpiryJson.Type == JTokenType.Null)
+            {
+                ExpiredTime = DateTime.MaxValue;
+            }
+            else
+            {
+                TokenExpiry tokenExpiry = tokenExpiryJson.ToObject<TokenExpiry>();
+
+                if (tokenExpiry == null)
+                {
+                    throw new InvalidDataException("Credential Json contains an invalid token_expiry.");
+                }
+
+                ExpiredTime = new DateTime(
+                    tokenExpiry.Year,
+                    tokenExpiry.Month,
+                    tokenExpiry.Day,
+                    tokenExpiry.Hour,
+                    tokenExpiry.Minute,
+                    tokenExpiry.Second,
+                    tokenExpiry.MicroSecond/1000,
+                    DateTimeKind.Utc);
+            }
+        }
 
         /// <summary>
         /// Returns <c>true</c> if the token is expired or it's going to be expired in the next minute
         /// or if the active user has changed.
         /// </summary>
-        public bool IsExpired(IClock clock)
+        public bool IsExpiredOrInvalid()
         {
-            if (AccessToken == null || !ExpiresInSeconds.HasValue)
+            if (AccessToken == null)
             {
                 return true;
             }
 
-            string currentAccount = CloudSdkSettings.GetSettingsValue("account");
+            string currentAccount = getActiveUser();
+            bool activeUserChanged = !string.Equals(currentAccount, activeUser);
 
-            return Issued.AddSeconds(ExpiresInSeconds.Value - 60) <= clock.Now
-                || !string.Equals(currentAccount, activeUser);
+            if (activeUserChanged)
+            {
+                return true;
+            }
+
+            if (ExpiredTime == DateTime.MaxValue)
+            {
+                return false;
+            }
+
+            return ExpiredTime.AddSeconds(-60) <= DateTime.UtcNow;
         }
     }
+
+    /// <summary>
+    /// Represents a token expiry object returned from gcloud auth print-access-token.
+    /// </summary>
+    internal class TokenExpiry
```

Comment:
This is only used as a shortcut when parsing the expiration from the JSON object, right? If so, why don't we make this a private inner-class of `ActiveUserToken`. The goal is to restrict the scope of this type name as much as possible, since it isn't used anywhere else.

Also, given the basic nature of the time, I'd personally be OK with removing the newlines between its properties and putting the attribute/properties on the same line to conserve space. e.g.

````
[JsonProperty("microsecond")] internal int MicroSecond { get; set; }
[JsonProperty("second")]      internal int Second { get; set; }
[JsonProperty("minute")]      internal int Minute { get; set; }
````

Just a matter of style. I'd suggest making the code as compact as possible, since it isn't as important.